### PR TITLE
(Re-)Enabling early stopping criterions when performing non-distributed training.

### PR DESCRIPTION
**Pull request by Lucaweihs**
A project I'm working on benefits from being able to run many small experiments in order to do hyperparameter tuning. Towards enabling HP tuning, this PR re-enables early stopping criterions (a feature we removed when we introduced distributed training) in the case where distributed training is not used.

### CODE REVIEW

File: allenact/algorithms/onpolicy_sync/losses/a2cacktr.py [hunk 1]
```diff
@@ -53,6 +53,11 @@ def loss_per_step(  # type: ignore
         )
 
         dist_entropy: torch.FloatTensor = actor_critic_output.distributions.entropy()
+        dist_entropy = dist_entropy.view(
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/losses/a2cacktr.py:56]: I'm not sure I understand the need to match the dimensions of entropy and log_probs (as far as I understand, both are reduced to a scalr before being added).

File: allenact/algorithms/onpolicy_sync/losses/ppo.py [hunk 2]
```diff
@@ -53,17 +53,23 @@ def loss_per_step(
...
 
+        dist_entropy: torch.FloatTensor = actor_critic_output.distributions.entropy()
+
+        def add_trailing_dims(t: torch.Tensor):
+            assert len(t.shape) <= len(batch["norm_adv_targ"].shape)
+            return t.view(
+                t.shape + ((1,) * (len(batch["norm_adv_targ"].shape) - len(t.shape)))
+            )
+
+        dist_entropy = add_trailing_dims(dist_entropy)
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/losses/ppo.py:67]: (Same comment as in A2C)

File: allenact/algorithms/onpolicy_sync/engine.py [hunk 3]
```diff
@@ -651,6 +658,18 @@ def __init__(
...
+        if self.num_workers != 1:
+            # Ensure that we're only using early stopping criterions in the non-distributed setting.
+            if any(
+                stage.early_stopping_criterion is not None
+                for stage in self.training_pipeline.pipeline_stages
+            ):
+                raise NotImplementedError(
+                    "Early stopping criterions are currently only allowed when using a single training worker, i.e."
+                    " no distributed (multi-GPU) training. If this is a feature you'd like please create an issue"
+                    " at https://github.com/allenai/allenact/issues."
```
Comment (jordis-ai2) [inline allenact/algorithms/onpolicy_sync/engine.py:670]: I would add "... or feel free to make a PR with your own implementation for us to review".

File: allenact/base_abstractions/sensor.py [hunk 4]
```diff
@@ -236,7 +236,7 @@ def __init__(
     ) -> None:
         self.nactions = nactions
         self.expert_args: Dict[str, Any] = expert_args or {}
-
+        observation_space = self._get_observation_space()
```
Comment (jordis-ai2) [inline allenact/base_abstractions/sensor.py:239]: This will change quite a bit with the hierarchical policies PR, but it was missing, yes.

File: allenact/utils/system.py [hunk 5]
```diff
@@ -85,6 +85,7 @@ def init_logging(human_log_level: str = "info") -> None:
     the log_level to one of `HUMAN_LOG_LEVELS`. And sets up a handler
     for stderr. The logging level is propagated to all subprocesses.
     """
+    human_log_level = human_log_level.lower().strip()
```
Comment (jordis-ai2) [inline allenact/utils/system.py:88]: I felt it 👍

File: allenact_plugins/lighthouse_plugin/lighthouse_tasks.py [hunk 6]
```diff
@@ -276,20 +284,27 @@ def query_expert(self, **kwargs) -> Tuple[Any, bool]:
...
             )
         else:
             raise NotImplementedError("Can only query expert for world dims of 1 or 2.")
 
+        if return_policy:
+            return policy, True
+        elif deterministic:
+            return int(np.argmax(policy)), True
+        else:
+            return (
```
Comment (jordis-ai2) [inline allenact_plugins/lighthouse_plugin/lighthouse_tasks.py:303]: Is this a noisy expert?

**Review by jordis-ai2 — COMMENTED** [review]
Just a couple of comments/questions, but other than that LGTM.

File: allenact/algorithms/onpolicy_sync/losses/a2cacktr.py (hunk 1 above)
Comment (Lucaweihs) [reply allenact/algorithms/onpolicy_sync/losses/a2cacktr.py:56]: This is mostly unnecessary, it's just that this `def loss_per_step(...)` function returns everything before doing reduction and if other losses want to use these pre-reduction outputs (e.g. ADVISOR-type losses), it's a bit easier to do so if the dimensions match.

File: allenact/base_abstractions/sensor.py (hunk 4 above)
Comment (Lucaweihs) [reply allenact/base_abstractions/sensor.py:239]: Yah, some of my code was complaining so I added this quick band aid, looking forward to your solution being merged :)

File: allenact_plugins/lighthouse_plugin/lighthouse_tasks.py (hunk 6 above)
Comment (Lucaweihs) [reply allenact_plugins/lighthouse_plugin/lighthouse_tasks.py:303]: It's an expert where several actions are equally good at certain timesteps and so, previously, I'd returned a policy that put equal weight on all of these policies. I'm trying to port some old code to use AllenAct >=0.3 and, as the expert policy loss fixes are still pending, I added this to allow for deterministically returning actions rather than policies.

File: allenact/algorithms/onpolicy_sync/losses/a2cacktr.py (hunk 1 above)
Comment (jordis-ai2) [reply allenact/algorithms/onpolicy_sync/losses/a2cacktr.py:56]: I agree.
